### DISCUSSION TITLE
Positioning of elements with @tstamp/@tstamp2 versus @startid/@endid

### DISCUSSION THREAD

**Discussion by wergo** [rism-digital/verovio#3313] [General]
Verovio renders control elements with @startid/@endid attributes differently than the same elements coded with `@tstamp`/`@tstamp2` attributes. While in case of slurs, the rendering with time stamps seems clearly incorrect, the behavior of the other elements is inconsistent, but it is not immediately evident, which of the renderings is better. I would be happy to your opinion on this.

The following example encodes several such elements with ids in the first staff, and with time stamps in the second: 

<img width="1395" alt="Bildschirm­foto 2023-03-07 um 13 53 30" src="https://user-images.githubusercontent.com/45632600/223432246-04f63519-e9ea-439a-96cf-dcbd6494acf3.png">

Example code here: https://mei-friend.mdw.ac.at/?file=https://raw.githubusercontent.com/wergo/test-encodings/main/slurs-ids-vs-tstamps.mei

1) `slur`: look correct with ids; but with tstamps both ends of the slur seem to be too far left and align with the left corner of the note heads; the right end of the slur seems to be too low in all four cases. (Version 3.7 had a basic rendering of time stamped slurs, since version 3.8 it looks like the present image, created with 3.15).

2) `hairpin`: the time stamp hairpin aligns with the left side of the note heads, while the start/endid version aligns with the note head center. In this case, the time stamp version seems better. However, if stem.dir is reversed (select notes and press keyboard shortcut `X` on mei-friend), it still aligns with the left side, which then looks less favorable.

3) `dynam`: are centered on the note heads with ids (seems correct), but aligned with the left side with time stamps (less favorable)

4) `mordent`: with ids centered (good), but shifted quite to the left with tstamp (less good)

5) `pedal`: again centered with ids (good), left aligned with tstamp (almost no difference here)

Related issues: #2522 and #2557

**Comment by lpugin** [comment]
First of all, a few remarks on how time stamps and ids work in Verovio. Time stamps in MEI are visual positions expressed in time signature units. When the time stamp corresponds to the position of a note, it is its left position because this is how notes are aligned in music notation. For example, if you have a quarter note and a whole note, this is how they have to be aligned:
![image](https://user-images.githubusercontent.com/689412/223453926-b1307cb7-6f26-46bd-b2b1-b1391a2654d6.png)
When using `startid` or `endid`, then the visual position can be (and is) adjusted according to the element pointed at. For example, for a dynamic, it can be centered on the note, also because this is what you would expect in music notation and because it would be weird not to do so:
![image](https://user-images.githubusercontent.com/689412/223455981-1f22e891-4b8f-4847-af46-1f81df67ee35.png)
This also means that the position will be different depending if the note pointed at is a quarter note or a whole note because the note head radius is different. That seems a normal behaviour.

So yes, time stamps and ids do work differently, and I do not think we can avoid that. Basically Verovio renders what the encoding expresses. I am not against changing things, of course, so if you have an ideas we can look at it. However, trying to infer the corresponding element a time stamp corresponds to is not a realistic option.

**Reply by wergo** [reply]
Thanks a lot for this great explanation. I completely understand that these are different ways of encoding and that they are handled differently by Verovio. I don't have a good suggestion how to change it, and it is therefore probably not necessary.

However, the slurs seem to be an exception. Their left ends actually do change as a function of the height of the note at that time stamp, while the right ends seem to always have the same y-value irrespective of the note there. This could be improved.

**Reply by lpugin** [reply]
Yes, maybe we can improve this. I will also add the explanation to the documentation.

I think one more fundamental question we could discuss is the use of both time stamps and ids. The assumption that has been made in Verovio is that we should give priority to ids because that are semantically better and because they allow for the placement to be adjusted according to the element the id is referring to. When both ids and time stamps are give, the latter are currently ignored in Verovio, simply yielding a warning. One reason for that is that we have (or used to have?) a lot of data converted with the musicxml2mei xslt that would always encode both ids and timestamps. With such data, it just appeared to be better to ignore ids.

Now in several cases, I thought we could see it differently, and basically do the opposite. That is, prioritising time stamps for visual placement, while the ids are used only for the semantic (e.g. in MIDI output). One typical example could be a slur, which you might want to be shorter, but still semantically attached to a note. (The other advantage here would be that the link to the note through the id would give us some information about the y position of the slur, which we currently do not have with time stamps only.)

Changing this might require it to be optional with an option, and we would need to think is it makes sense to change it for all control event, or category by category (slur, dynam, hairpin), or even case by case, which then would require something to be given in the encoding.
